core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -250,10 +250,11 @@ public boolean equals(Object o) {
 
         ByteSizeValue sizeValue = (ByteSizeValue) o;
 
-        if (size != sizeValue.size) return false;
-        if (sizeUnit != sizeValue.sizeUnit) return false;
-
-        return true;
+        if (sizeUnit == sizeValue.sizeUnit) {
+            return size == sizeValue.size;
+        } else {
+            return bytes() == sizeValue.bytes();
+        }
```

Comment:
I tried to avoid conversions in the case when it used to return true (i.e., same units) in order to have the minimal change for performance. Can switch if you feel strongly about it.